File: 2003-distinct-numbers-in-each-subarray/distinct-numbers-in-each-subarray.py
class Solution:
    def distinctNumbers(self, nums: List[int], k: int) -> List[int]:
        d={}
        ans=[]
        for i in range(k):
            d[nums[i]]=d.get(nums[i],0)+1
        ans.append(len(d))
        for i in range(k,len(nums)):
            d[nums[i-k]]=d.get(nums[i-k],0)-1
            if d[nums[i-k]]==0:
                del d[nums[i-k]]
            d[nums[i]]=d.get(nums[i],0)+1
            ans.append(len(d))
        return ans
        

File: 2003-distinct-numbers-in-each-subarray/distinct-numbers-in-each-subarray.java
class Solution {
    public int[] distinctNumbers(int[] nums, int k) {
        ArrayList<Integer> ans=new ArrayList<>();
        HashMap<Integer,Integer> map=new HashMap<>();
        for(int i=0;i<k;i++){
            map.put(nums[i],map.getOrDefault(nums[i],0)+1);
        }
        ans.add(map.size());
        for(int i=k;i<nums.length;i++){
            map.put(nums[i-k],map.get(nums[i-k])-1);
            if(map.get(nums[i-k])==0){
                map.remove(nums[i-k]);
            }
            map.put(nums[i],map.getOrDefault(nums[i],0)+1);
            ans.add(map.size());
        }
        int size=ans.size();
        int []finalAnswer=new int[size];
        for(int i=0;i<size;i++){
            finalAnswer[i]=ans.get(i);
        }
        return finalAnswer;
    }
}